crates/lib/src/execution_data.rs

```diff
@@ -1,58 +1,121 @@
+    /// TODO
+    Complex32(Complex32),
+    /// TODO
+    Complex64(Complex64),
+}
+
+/// A matrix where rows are the values of a memory index across all shots, and columns are values
+/// for all memory indexes of a given shot.
+pub type RegisterMatrix = Array2<Option<ReadoutValue>>;
+
 #[derive(Debug, Clone, PartialEq)]
 #[repr(transparent)]
-pub struct ReadoutMap(HashMap<MemoryReference, ReadoutValues>);
+pub struct ReadoutMap(HashMap<String, RegisterMatrix>);
 
 impl ReadoutMap {
-    /// Given a known readout field name and index, return the result's [`ReadoutValues`], if any.
-    pub fn get_readout_values(&self, field: String, index: u64) -> Option<ReadoutValues> {
-        let readout_values = self.0.get(&MemoryReference { name: field, index })?;
+    /// Returns the individual value for the given memory index and shot number, or None, if it
+    /// doesn't exist
+    pub fn get_value(
+        &self,
+        register_name: &str,
+        index: usize,
+        shot_num: usize,
+    ) -> Option<ReadoutValue> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if index > matrix.nrows() || shot_num > matrix.ncols() {
+                return None;
+            }
+            matrix[[index, shot_num]]
+        } else {
+            None
+        }
+    }
 
-        Some(readout_values.clone())
+    /// Returns a vector containing all the values in the given register at a particular memory
+    /// index across all shots.
+    pub fn get_values_by_memory_index(
+        &self,
+        register_name: &str,
+        index: usize,
+    ) -> Option<Vec<Option<ReadoutValue>>> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if index > matrix.nrows() {
+                return None;
+            }
+            Some(matrix.row(index).to_vec())
+        } else {
+            None
+        }
     }
 
-    /// Given a known readout field name, return the result's [`ReadoutValues`] for all indices, if any.
-    pub fn get_readout_values_for_field(
+    /// Returns a vector containing all the values in the given register for a particular shot number.
+    pub fn get_values_by_shot(
         &self,
-        field: &str,
-    ) -> Result<Option<Vec<Option<ReadoutValues>>>, TryFromIntError> {
-        let mut readout_values = Vec::new();
-        for (memref, values) in &self.0 {
-            let MemoryReference { name, index } = memref;
-            let index = usize::try_from(*index)?;
-            if name == field {
-                if readout_values.len() <= index {
-                    readout_values.resize(index + 1, None);
-                }
-                readout_values[index] = Some(values.clone());
+        register_name: &str,
+        shot: usize,
+    ) -> Option<Vec<Option<ReadoutValue>>> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if shot > matrix.ncols() {
```

Comment:
Same as rows: maybe I missed something, seems like you should be able to avoid bounds checking.

```suggestion
            if shot >= matrix.ncols() {
```

Reply:
On an `ndarray` `get` works like it does on a vec, so we are able to get away without bounds checking because it'll just return None if the index is out of bounds. The `row` and `column` methods will panic if the index is out of range, so I think we have to check the bounds - but I could be missing something, I would love to avoid having to check it if possible.